File: Leetcoding-Actions/LC-Monthly-Challenges/2020-11/20-Search-in-Rotated-Sorted-Array-ii.py
class Solution:
    def search(self, nums: List[int], target: int) -> bool:
        '''
        T: O(n) worst; O(log n) best and S: O(1)
        '''
        lo, hi = 0, len(nums) - 1
        while lo <= hi:
            mid = lo + (hi - lo) // 2
            
            if nums[mid] == target:
                return True
            
            elif nums[lo] == nums[mid] == nums[hi]:
                lo += 1
                hi -= 1
                
            elif nums[lo] <= nums[mid]: # Left is sorted
                if nums[lo] <= target < nums[mid]:
                    hi = mid - 1
                else:
                    lo = mid + 1
                    
            else: # right is sorted
                if nums[mid] < target <= nums[hi]:
                    lo = mid + 1
                else:
                    hi = mid - 1
        return False

File: Leetcoding-Actions/LC-Monthly-Challenges/2020-11/20-Search-in-Rotated-Sorted-Array-ii.java
class Solution {
    public boolean search(int[] nums, int target) {
        /**
        * T: O(n) worst; O(log n) best and S: O(1)
        */
        int n = nums.length;
        int lo = 0;
        int hi = n - 1;
        
        while (lo <= hi) {
            int mid = lo + (hi - lo) / 2;
            
            if (nums[mid] == target) {
                return true;
            } else if (nums[lo] == nums[mid] && nums[mid] == nums[hi]) {
                lo++;
                hi--;
            } else if (nums[lo] <= nums[mid]) {
                if (nums[lo] <= target && target < nums[mid]) {
                    hi = mid - 1;
                } else {
                    lo = mid + 1;
                }
            } else {
                if (nums[mid] < target && target <= nums[hi]) {
                    lo = mid + 1;
                } else {
                    hi = mid - 1;
                }
            }
        }
        return false;
    }
}
